sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -487,6 +487,13 @@ static void _process_coap_pdu(gcoap_socket_t *sock, sock_udp_ep_t *remote, sock_
         }
         else {
             DEBUG("gcoap: msg not found for ID: %u\n", coap_get_id(&pdu));
+            if (coap_get_type(&pdu) == COAP_TYPE_CON) {
+                /* we might run into this if an ACK to a sender got lost
+                 * => best to ACK (again) */
+                messagelayer_emptyresponse_type = COAP_TYPE_ACK;
+                DEBUG("gcoap: Answering unknown CON response with ACK to "
+                      "shut up sender\n");
+            }
```

Comment:
A reset should be send:

>   In case a message carrying a response is unexpected (the client is
>   not waiting for a response from the identified endpoint, at the
>   endpoint addressed, and/or with the given token), the response is
>   rejected (Sections 4.2 and 4.3).
>
>   Implementation Note:  A client that receives a response in a CON
>      message may want to clean up the message state right after sending
>      the ACK.  If that ACK is lost and the server retransmits the CON,
>      the client may no longer have any state to which to correlate this
>      response, making the retransmission an unexpected message; the
>      client will likely send a Reset message so it does not receive any
>      more retransmissions.  This behavior is normal and not an
>      indication of an error.  (Clients that are not aggressively
>      optimized in their state memory usage will still have message
>      state that will identify the second CON as a retransmission.
>      Clients that actually expect more messages from the server
>      [OBSERVE] will have to keep state in any case.)

https://datatracker.ietf.org/doc/html/rfc7252#section-5.3.2

Reply:
I agree, an RST is the intended behavior. (I thought we already did this, at least that's what I remember of last time touching the empty-ACK code).

Reply:
> I agree, an RST is the intended behavior. (I thought we already did this, at least that's what I remember of last time touching the empty-ACK code).

I saw a server keep sending transmissions even though the CON was already ACK'd. I did not see any RST messages.

If the RFC says RST should be used, I use RST.

Reply:
Yeah, I just failed to pay attention to the "else /* no memo */" branch, thanks for the fix.